File: HealthyPlant/HealthyPlant/Models/PlantProfile.swift
import Foundation

struct PlantProfile: Codable, Identifiable, Equatable, Hashable {
    let id: String
    let userId: String
    var name: String
    var plantType: String
    var photoURL: String?
    var ageDays: Int
    var plantedDate: String
    var heightFeet: Int
    var heightInches: Int
    var sunNeeds: String?
    var waterNeeds: String?
    var harvestTime: String?
    var wateringFrequencyDays: Int?
    var sunHoursMin: Int?
    var sunHoursMax: Int?
    var aiLastUpdated: String?
    let createdAt: String
    var updatedAt: String

    /// Formatted height string, e.g. "2ft 6in"
    var formattedHeight: String {
        Int.formatHeight(feet: heightFeet, inches: heightInches)
    }

    /// Formatted age string, e.g. "1yr 3mo 5d"
    var formattedAge: String {
        let years = ageDays / 365
        let months = (ageDays % 365) / 30
        let days = ageDays % 30
        var parts: [String] = []
        if years > 0 { parts.append("\(years)yr") }
        if months > 0 { parts.append("\(months)mo") }
        if days > 0 || parts.isEmpty { parts.append("\(days)d") }
        return parts.joined(separator: " ")
    }

    // MARK: - Preview Mock

    static let mock = PlantProfile(
        id: "mock-001",
        userId: "user-001",
        name: "Tommy Tomato",
        plantType: "Tomato",
        photoURL: nil,
        ageDays: 45,
        plantedDate: "2026-02-05",
        heightFeet: 1,
        heightInches: 8,
        sunNeeds: "Full sun (6-8 hours)",
        waterNeeds: "Water every 2-3 days",
        harvestTime: "60-80 days from transplant",
        wateringFrequencyDays: 3,
        sunHoursMin: 6,
        sunHoursMax: 8,
        aiLastUpdated: "2026-03-22T00:00:00+00:00",
        createdAt: "2026-02-05T00:00:00+00:00",
        updatedAt: "2026-03-22T00:00:00+00:00"
    )

    static let mockList: [PlantProfile] = [
        mock,
        PlantProfile(
            id: "mock-002",
            userId: "user-001",
            name: "Basil Buddy",
            plantType: "Basil",
            photoURL: nil,
            ageDays: 20,
            plantedDate: "2026-03-02",
            heightFeet: 0,
            heightInches: 10,
            sunNeeds: "Partial to full sun",
            waterNeeds: "Keep soil moist",
            harvestTime: "3-4 weeks for first harvest",
            wateringFrequencyDays: nil,
            sunHoursMin: nil,
            sunHoursMax: nil,
            aiLastUpdated: "2026-03-22T00:00:00+00:00",
            createdAt: "2026-03-02T00:00:00+00:00",
            updatedAt: "2026-03-22T00:00:00+00:00"
        ),
    ]
}

File: backend/api/models/plant_profile.py
"""Pydantic models for plant profiles."""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class PlantProfileCreate(BaseModel):
    """Request model for creating a plant profile."""

    name: str = Field(..., min_length=1, max_length=100, description="Display name for the plant")
    plant_type: str = Field("", alias="plantType", max_length=100, description="Type/species of the plant")
    photo_url: Optional[str] = Field(None, alias="photoURL", description="URL of the plant photo")
    age_days: int = Field(..., alias="ageDays", ge=0, description="Age of the plant in days")
    planted_date: str = Field(..., alias="plantedDate", description="Date the plant was planted (YYYY-MM-DD)")
    height_feet: int = Field(..., alias="heightFeet", ge=0, description="Height in feet")
    height_inches: int = Field(..., alias="heightInches", ge=0, le=11, description="Height remaining inches")

    model_config = {"populate_by_name": True}


class PlantProfileUpdate(BaseModel):
    """Request model for updating a plant profile. All fields optional."""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    plant_type: Optional[str] = Field(None, alias="plantType", min_length=1, max_length=100)
    photo_url: Optional[str] = Field(None, alias="photoURL")
    age_days: Optional[int] = Field(None, alias="ageDays", ge=0)
    planted_date: Optional[str] = Field(None, alias="plantedDate")
    height_feet: Optional[int] = Field(None, alias="heightFeet", ge=0)
    height_inches: Optional[int] = Field(None, alias="heightInches", ge=0, le=11)
    sun_needs: Optional[str] = Field(None, alias="sunNeeds", description="AI-generated sun requirements")
    water_needs: Optional[str] = Field(None, alias="waterNeeds", description="AI-generated water requirements")
    harvest_time: Optional[str] = Field(None, alias="harvestTime", description="AI-generated harvest time estimate")

    model_config = {"populate_by_name": True}


class PlantProfileResponse(BaseModel):
    """Response model for a plant profile."""

    id: str = Field(..., description="Firestore document ID")
    user_id: str = Field(..., alias="userId")
    name: str
    plant_type: str = Field(..., alias="plantType")
    photo_url: Optional[str] = Field(None, alias="photoURL")
    age_days: int = Field(..., alias="ageDays")
    planted_date: str = Field(..., alias="plantedDate")
    height_feet: int = Field(..., alias="heightFeet")
    height_inches: int = Field(..., alias="heightInches")
    sun_needs: Optional[str] = Field(None, alias="sunNeeds")
    water_needs: Optional[str] = Field(None, alias="waterNeeds")
    harvest_time: Optional[str] = Field(None, alias="harvestTime")
    watering_frequency_days: Optional[int] = Field(None, alias="wateringFrequencyDays")
    sun_hours_min: Optional[int] = Field(None, alias="sunHoursMin")
    sun_hours_max: Optional[int] = Field(None, alias="sunHoursMax")
    ai_last_updated: Optional[str] = Field(None, alias="aiLastUpdated")
    sensor_id: Optional[str] = Field(None, alias="sensorId")
    sensor_last_reading: Optional[Any] = Field(None, alias="sensorLastReading")
    created_at: str = Field(..., alias="createdAt")
    updated_at: str = Field(..., alias="updatedAt")

    model_config = {"populate_by_name": True, "from_attributes": True}
